crates/fs/private/previous.rs

```diff
@@ -0,0 +1,843 @@
+    discrepancy_budget: usize,
+    /// The history of each path segment leading up to the final node
+    path: Vec<PathSegmentHistory>,
+    /// The target node's history
+    target: PrivateNodeHistory,
+}
+
+struct PathSegmentHistory {
+    /// The directory that the history was originally created relative to.
+    dir: Rc<PrivateDirectory>,
+    /// The history of said directory.
+    history: PrivateNodeHistory,
+    /// The name of the child node to follow for history next.
+    path_segment: String,
+}
+
+/// This represents the state of an iterator through the history of
+/// only a single private node. It can only be constructed when you
+/// know the past ratchet state of such a node.
+pub struct PrivateNodeHistory {
+    /// Keep a reference to the version of the forest used upon construction.
+    /// It could *technically* change what's behind a certain key in between
+    /// previous node requests, this forces it to be consistent.
+    forest: Rc<PrivateForest>,
+    /// The private node header is all we need to look up private nodes in the forest.
+    /// This will always be the header of the *next* version after what's retrieved from
+    /// the `ratchets` iterator.
+    header: PrivateNodeHeader,
+    /// The iterator for previous revision ratchets.
+    ratchets: PreviousIterator,
+}
+
+impl PrivateNodeHistory {
+    /// Create a history iterator for given private node up until `past_ratchet`.
+    ///
+    /// There must be an `n > 0` for which `node.get_header().ratchet == past_ratchet.inc_by(n)`.
+    ///
+    /// Discrepancy budget is used to bound the search for the actual `n`
+    /// and prevent infinite looping in case it doesn't exist.
+    pub fn of(
+        node: &PrivateNode,
+        past_ratchet: &Ratchet,
+        discrepancy_budget: usize,
+        forest: Rc<PrivateForest>,
+    ) -> Result<Self> {
+        Self::from_header(
+            node.get_header().clone(),
+            past_ratchet,
+            discrepancy_budget,
+            forest,
+        )
+    }
+
+    /// Create a history iterator for a node given its header.
+    ///
+    /// See also `PrivateNodeHistory::of`.
+    pub fn from_header(
+        header: PrivateNodeHeader,
+        past_ratchet: &Ratchet,
+        discrepancy_budget: usize,
+        forest: Rc<PrivateForest>,
+    ) -> Result<Self> {
+        let forest = Rc::clone(&forest);
+        let ratchets = PreviousIterator::new(past_ratchet, &header.ratchet, discrepancy_budget)
+            .map_err(FsError::PreviousError)?;
+        Ok(PrivateNodeHistory {
+            forest,
+            header,
+            ratchets,
+        })
+    }
+
+    /// Step the history one step back and retrieve the private node at the
+    /// previous point in history.
+    ///
+    /// Returns `None` if there is no such node in the `PrivateForest` at that point in time.
+    pub async fn previous_node<B: BlockStore>(&mut self, store: &B) -> Result<Option<PrivateNode>> {
```

Comment:
Anything that works with a `BlockStore` could potentially not find the required blocks. I think there's a difference between "there is no more previous entry" and "something went wrong when trying to figure out what the previous entry was".

Reply:
@matheus23 yeah, I do get that; however, in this case, get_private_ref has nothing that would bail! on it, and therefore there would not need to be a result here at all. I tend to think of results when we know this will absolutely go wrong/crash, but I'm not seeing that in this example, seemingly?

Reply:
Ok, followed the tree outside of the PR, and see where the `get` call leads to a result. To this point then, maybe `get_private_ref` doesn't need it. In terms of naming then, I may prefer `get_previous_node` which seems to favor a resulting type.